### PR TITLE
add event emissions to provider and consumer

### PR DESCRIPTION
**Pull request by MSalopek**
Events added:
* `create_client` (provider -> during CreateConsumerClient call)
* `consumer_added` (provider)
* `execute_consumer_chain_slash` (provider)
* `fee_channel_opened` (consumer)
* `first_vsc_packet` (consumer -> mainly for debugging so we know that first vsc was received)
* `fee_distribution` (consumer)
* `send_slash_packet` (consumer)
* `send_matured_vsc_packet` (consumer)

### CODE REVIEW

File: x/ccv/consumer/module.go [hunk 1]
```diff
@@ -170,6 +173,9 @@ func (am AppModule) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
 	am.keeper.TrackHistoricalInfo(ctx)
 }
 
+// TODO: add events
+// - provider fee distribution transmission
+// - matured validator updates (the consumer could emit in keeper.EndBlockVSU)
```
Comment (MSalopek) [inline x/ccv/consumer/module.go:178]: And other validator stuff

File: x/ccv/democracy/governance/module.go [hunk 2]
```diff
@@ -41,6 +41,8 @@ func NewAppModule(cdc codec.Codec, keeper keeper.Keeper, ak govtypes.AccountKeep
 	}
 }
 
+// TODO: add events
+// - maybe emit ids of forbidden proposals
```
Comment (MSalopek) [inline x/ccv/democracy/governance/module.go:45]: Skip for now

File: x/ccv/provider/keeper/proposal.go [hunk 3]
```diff
@@ -94,6 +95,20 @@ func (k Keeper) CreateConsumerClient(ctx sdk.Context, chainID string,
 	if lockUbdOnTimeout {
 		k.SetLockUnbondingOnTimeout(ctx, chainID)
 	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			clienttypes.EventTypeCreateClient,
```
Comment (mpoke) [inline x/ccv/provider/keeper/proposal.go:101]: Change this event to `EventTypeConsumerClientCreated`.

File: x/ccv/consumer/keeper/relay.go [hunk 4]
```diff
@@ -36,6 +37,18 @@ func (k Keeper) OnRecvVSCPacket(ctx sdk.Context, packet channeltypes.Packet, new
 		k.SetProviderChannel(ctx, packet.DestinationChannel)
 		// - send pending slash requests in states
 		k.SendPendingSlashRequests(ctx)
+
+		// emit first VSC packet to signal that CCV is working
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				ccv.EventTypeFirstVSCPacket,
```
Comment (mpoke) [inline x/ccv/consumer/keeper/relay.go:44]: Change this event to `EventTypeCCVChannelEstablished`.

File: x/ccv/provider/keeper/keeper.go [hunk 5]
```diff
@@ -293,6 +293,17 @@ func (k Keeper) SetConsumerChain(ctx sdk.Context, channelID string) error {
 	k.SetInitChainHeight(ctx, chainID, uint64(ctx.BlockHeight()))
 	// - remove init timeout timestamp
 	k.DeleteInitTimeoutTimestamp(ctx, chainID)
+
+	// emit event on successful addition
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			ccv.EventTypeConsumerChainAdded,
```
Comment (mpoke) [inline x/ccv/provider/keeper/keeper.go:300]: Change this event to `EventTypeCCVChannelEstablished`.

File: x/ccv/consumer/keeper/relay.go [hunk 6]
```diff
@@ -36,6 +37,18 @@ func (k Keeper) OnRecvVSCPacket(ctx sdk.Context, packet channeltypes.Packet, new
...
+				sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+				sdk.NewAttribute(channeltypes.AttributeKeySrcChannel, packet.SourceChannel),
+				sdk.NewAttribute(channeltypes.AttributeKeySrcPort, packet.SourcePort),
+				sdk.NewAttribute(channeltypes.AttributeKeyDstChannel, packet.DestinationChannel),
+				sdk.NewAttribute(channeltypes.AttributeKeyDstPort, packet.DestinationPort),
```
Comment (mpoke) [inline x/ccv/consumer/keeper/relay.go:49]: I think we don't need all these fields. Only the channelID and portID on the consumer side.

File: x/ccv/consumer/keeper/relay.go [hunk 7]
```diff
@@ -110,6 +123,16 @@ func (k Keeper) SendVSCMaturedPackets(ctx sdk.Context) error {
 				return err
 			}
 			k.DeletePacketMaturityTime(ctx, vscId)
+			ctx.EventManager().EmitEvent(
+				sdk.NewEvent(
+					ccv.EventTypeSendMaturedVSCPacket,
+					sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+					sdk.NewAttribute(ccv.AttributeChainID, ctx.ChainID()),
```
Comment (mpoke) [inline x/ccv/consumer/keeper/relay.go:130]: I don't think we need the chain ID given that the user already needs to subscribe to the chain to get this event.

File: x/ccv/consumer/keeper/relay.go (hunk 6 above)
Comment (MSalopek) [reply x/ccv/consumer/keeper/relay.go]: `Done`

File: x/ccv/consumer/keeper/relay.go (hunk 7 above)
Comment (MSalopek) [reply x/ccv/consumer/keeper/relay.go:128]: Having a chainID visible could be useful.

The user could be listening to events from multiple consumer chains in parallel and merge them into a single processing pipeline

In that case it is alot easier to manage if you already have a chain ID

File: x/ccv/provider/keeper/keeper.go (hunk 5 above)
Comment (MSalopek) [reply x/ccv/provider/keeper/keeper.go]: `Done`.

File: x/ccv/types/events.go [hunk 8]
```diff
@@ -2,11 +2,38 @@ package types
...
-	EventTypePacket       = "ccv_packet"
-	EventTypeChannelClose = "channel_closed"
+	EventTypeTimeout                  = "timeout"
+	EventTypePacket                   = "ccv_packet"
+	EventTypeChannelClose             = "channel_closed"
+	EventTypeChannelEstablished       = "channel_established"
+	EventTypeFeeTransferChannelOpened = "fee_channel_opened"
+	EventTypeConsumerClientCreated    = "consumer_client_created"
+
+	EventExecuteConsumerChainSlash = "execute_consumer_chain_slash"
```
Comment (danwt) [inline x/ccv/types/events.go:12]: Should line 12 have `EventType` as a prefix?

File: x/ccv/types/events.go [hunk 9]
```diff
@@ -2,11 +2,38 @@ package types
...
+	EventTypeFeeDistribution       = "fee_distribution"
+	EventTypeSendSlashPacket       = "send_slash_packet"
+	EventTypeSendMaturedVSCPacket  = "send_matured_vsc_packet"
```
Comment (danwt) [inline x/ccv/types/events.go:15]: Some of these are in past tense and some are in present tense, could it be more consistent?

By the way there is a nice vscode extension `change-case` which lets you easily change Camel to snake ect. Some of these dont match (ChannelClose = "channel_closed")

**Review by danwt — APPROVED** [review]
Nice work! I don't see a reason not to approve as this clearly cannot introduce bugs. Improvements and more events can be added later if necessary.

File: x/ccv/types/events.go (hunk 8 above)
Comment (MSalopek) [reply x/ccv/types/events.go]: Added.

File: x/ccv/types/events.go (hunk 9 above)
Comment (MSalopek) [reply x/ccv/types/events.go]: Changed cases that did not match.

File: x/ccv/provider/keeper/proposal.go (hunk 3 above)
Comment (smarshall-spitzbart) [inline x/ccv/provider/keeper/proposal.go:101]: This should reference the new `EventTypeConsumerClientCreated` type, correct?

**Review by smarshall-spitzbart — COMMENTED** [review]
Just a couple small things but great stuff!

File: x/ccv/types/events.go [hunk 10]
```diff
@@ -2,11 +2,38 @@ package types
 
 // CCV events
 const (
-	EventTypeTimeout      = "timeout"
-	EventTypePacket       = "ccv_packet"
-	EventTypeChannelClose = "channel_closed"
+	EventTypeTimeout                  = "timeout"
+	EventTypePacket                   = "ccv_packet"
+	EventTypeChannelClosed            = "channel_closed"
```
[the remark this answers is not in the document]
Comment (MSalopek) [reply x/ccv/types/events.go]: Removed.

File: x/ccv/provider/keeper/proposal.go (hunk 3 above)
Comment (MSalopek) [reply x/ccv/provider/keeper/proposal.go]: Changed.

**Review by smarshall-spitzbart — APPROVED** [review]
LGTM

**Pull request merged**